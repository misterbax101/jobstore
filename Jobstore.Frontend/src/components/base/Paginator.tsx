import React, { useState } from 'react';
import { Pagination, PaginationItem, PaginationLink } from 'reactstrap';

import { getPagesCount } from './../../untils/helper';


interface PaginatorProps {
    currentPage?: number,
    itemsCount: number,
    pageSize?: number,
    className?: string
    onPageChange: (newPage: number) => void;
}

const Paginator: React.FC<PaginatorProps> = ({ currentPage = 1, pageSize = 5, itemsCount, onPageChange, className }) => {

    const [selectedPage, setSelectedPage] = useState<number>(currentPage);

    const pagesCount = getPagesCount(itemsCount, pageSize);

    const goToPage = (page: number): void => {
        setSelectedPage(page);
        onPageChange(page);
    }

    const goToPreviousPage = (): void => {
        if (selectedPage !== 1) {
            goToPage(selectedPage - 1);
        }
    }

    const goToNextPage = (): void => {
        if (selectedPage !== pagesCount) {
            goToPage(selectedPage + 1);
        }
    }

    const isPageSelected = (page: number): boolean => page === selectedPage;

    const renderPageItems = (): JSX.Element[] => {
        const resut: JSX.Element[] = [];
        for (let page = 1; page <= pagesCount; ++page) {
            resut.push(
                <PaginationItem
                    key={page}
                    active={isPageSelected(page)}
                    onClick={() => goToPage(page)}>
                    <PaginationLink>
                        {page}
                    </PaginationLink>
                </PaginationItem>);
        }
        return resut;
    }
    return ((pagesCount <= 1) ? null :
        (<Pagination listClassName={className}>
            <PaginationItem
                disabled={isPageSelected(1)}
                onClick={goToPreviousPage}>
                <PaginationLink previous />
            </PaginationItem>
            {renderPageItems()}
            <PaginationItem
                onClick={goToNextPage}
                disabled={isPageSelected(pagesCount)}>
                <PaginationLink next />
            </PaginationItem>
        </Pagination>));
}

export default Paginator;
