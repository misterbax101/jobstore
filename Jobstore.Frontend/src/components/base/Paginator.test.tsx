import React from 'react';
import { shallow, mount } from 'enzyme';
import { Pagination } from 'reactstrap';

import Paginator, { PaginatorProps } from './Paginator';

const defaultProps: PaginatorProps =
{
    currentPage: 1,
    itemsCount: 15,
    pageSize: 5,
    onPageChange: (page: number) => { }
}

describe('<Paginator />', () => {
    it('should render correctly', () => {

        let wrapper = shallow(<Paginator {...defaultProps} />)

        expect(wrapper.getElements()).toMatchSnapshot();
    })

    it('should has class', () => {

        const className = 'test';
        let wrapper = mount(
            <Paginator
                {...defaultProps}
                className={className} />);

        expect(wrapper.find('ul').hasClass(className)).toBeTruthy();
    })

    it('should not render', () => {

        let wrapper = mount(
            <Paginator
                {...defaultProps} itemsCount={0} />);

        expect(wrapper.find('ul').length).toBe(0);
    })

    it('should render 3 pages', () => {

        let wrapper = mount(
            <Paginator
                {...defaultProps} itemsCount={3} pageSize={1} />);

        expect(wrapper.find('li').length).toBe(5);
    })

    it('should select second page', () => {

        let wrapper = mount(
            <Paginator
                {...defaultProps}
                itemsCount={3}
                pageSize={1}
                currentPage={2} />);

        expect(wrapper.find('li').at(2).hasClass('active')).toBeTruthy();
    })

    it('prev page should disabled', () => {

        let wrapper = mount(
            <Paginator
                {...defaultProps}
                itemsCount={3}
                pageSize={1}
                currentPage={1} />);

        expect(wrapper.find('li').first().hasClass('disabled')).toBeTruthy();
    })

    it('next page should disabled', () => {

        let wrapper = mount(
            <Paginator
                {...defaultProps}
                itemsCount={3}
                pageSize={1}
                currentPage={3} />);

        expect(wrapper.find('li').last().hasClass('disabled')).toBeTruthy();
    })

    it('should go to next page', () => {

        const onPageChangeMock = jest.fn((page: number) => {});
        let wrapper = mount(
            <Paginator
                {...defaultProps}
                itemsCount={3}
                pageSize={1}
                currentPage={1}
                onPageChange={onPageChangeMock} />);

        const pageItem = wrapper
            .find('li')
            .last();

        pageItem.simulate('click');

        expect(onPageChangeMock).toBeCalled();
        expect(onPageChangeMock).toBeCalledWith(2);
        expect(wrapper.find('li').at(2).hasClass('active')).toBeTruthy();
    })
})
