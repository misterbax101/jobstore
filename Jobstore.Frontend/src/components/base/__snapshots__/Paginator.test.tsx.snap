// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> should has class 1`] = `
Array [
  <Pagination
    aria-label="pagination"
    listClassName="test"
    listTag="ul"
    tag="nav"
  >
    <PaginationItem
      disabled={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        previous={true}
        tag="a"
      />
    </PaginationItem>
    <PaginationItem
      active={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        1
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        2
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        3
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      disabled={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        next={true}
        tag="a"
      />
    </PaginationItem>
  </Pagination>,
]
`;

exports[`<Paginator /> should render correctly 1`] = `
Array [
  <Pagination
    aria-label="pagination"
    listTag="ul"
    tag="nav"
  >
    <PaginationItem
      disabled={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        previous={true}
        tag="a"
      />
    </PaginationItem>
    <PaginationItem
      active={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        1
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        2
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        3
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      disabled={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        next={true}
        tag="a"
      />
    </PaginationItem>
  </Pagination>,
]
`;

exports[`<Paginator /> should renders 1`] = `
Array [
  <Pagination
    aria-label="pagination"
    listTag="ul"
    tag="nav"
  >
    <PaginationItem
      disabled={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        previous={true}
        tag="a"
      />
    </PaginationItem>
    <PaginationItem
      active={true}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        1
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        2
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      active={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        tag="a"
      >
        3
      </PaginationLink>
    </PaginationItem>
    <PaginationItem
      disabled={false}
      onClick={[Function]}
      tag="li"
    >
      <PaginationLink
        next={true}
        tag="a"
      />
    </PaginationItem>
  </Pagination>,
]
`;
